internal/contour/handler.go

```diff
@@ -164,11 +181,28 @@ func (e *EventHandler) Start(ctx context.Context) error {
 				// not to process it.
 				e.incSequence()
 			}
+
+			// We're done processing this event
+			if updateOpAdd, ok := op.(opAdd); ok {
+				if updateOpAdd.isInInitialList {
+					e.syncTracker.Finished()
+				}
+			}
 		case <-pending:
+			// Ensure informer caches are synced.
+			// Schedule a retry for dag rebuild if cache is not synced yet.
+			// Note that we can't block and wait for the cache sync as it depends on progress of this loop.
+			if !e.syncTracker.HasSynced() {
+				e.Info("skipping dag rebuild as cache is not synced")
+				timer.Reset(e.holdoffDelay)
+				break
+			}
+
```

Comment:
Do you suggest relocating the entire `if` block to line 189?

Reply:
Let me check this with you before committing:
```go
case op := <-e.update:
	if e.onUpdate(op) {
		outstanding++
		// If there is already a timer running, stop it.
		if timer != nil {
			timer.Stop()
		}

		// Schedule a DAG rebuild with holdoff delay
		if e.initialDagBuilt {
			delay := e.holdoffDelay
			if time.Since(lastDAGRebuild) > e.holdoffMaxDelay {
				// the maximum holdoff delay has been exceeded so schedule the update
				// immediately by delaying for 0ns.
				delay = 0
			}
			timer = time.NewTimer(delay)
			pending = timer.C
		}
	} else {
		// notify any watchers that we received the event but chose
		// not to process it.
		e.incSequence()
	}

	// We're done processing this event
	if updateOpAdd, ok := op.(opAdd); ok {
		if updateOpAdd.isInInitialList {
			e.syncTracker.Finished()
			// Schedule a DAG rebuild if the informer caches are synced
			if e.syncTracker.HasSynced() {
				timer = time.NewTimer(0)
			}
		}
	}
case <-pending:
	e.WithField("last_update", time.Since(lastDAGRebuild)).WithField("outstanding", reset()).Info("performing delayed update")
	e.rebuildDAG()
	e.incSequence()
	lastDAGRebuild = time.Now()
	e.initialDagBuilt = true
```

Reply:
Yes exactly.

Maybe consider small clarifications if following makes sense to you? 

```diff
-// Schedule a DAG rebuild with holdoff delay
+// Schedule a DAG rebuild with holdoff delay if initial DAG build has been done.

-// Schedule a DAG rebuild if the informer caches are synced
+// Schedule first DAG build after the informer caches are synced.
```

And I think `timer.Stop()` can be moved inside the if statement

```go
if e.initialDagBuilt {
    timer.Stop()
```
There it cannot be `nil` there anymore so the guard can be removed.

Reply:
Thanks! I've applied your suggestions.

Reply:
Hmm not sure about the CI test failures, but when I run this on my local machine, it will get stuck once every 10-30 runs. It will not print `scheduling the initial DAG build as caches are synced` (ever).

I thought that `HasSynced()` would turn from `false` to `true` only after last `Finished()` call, and therefore polling `HasSynced()` is unnecessary, but there must be something more in this?

With the previous approach where `HasSynced()` was polled periodically, I did not see it get stuck, though I'm uncertain if it is about polling or just because it was called ~100msec later. If I run the current version (without polling) under debugger, it does not seem to get stuck either, but it is much slower than without debugger.

Reply:
I think this is the cause:

Imagine the final object of the initial list is handled but one of the upstream `HasSynced` functions returns false. Then the `if e.syncTracker.HasSynced()` block will be skipped and will never have the chance to run.

Even if we move the block out of the `if updateOpAdd.isInInitialList` block, it may again occur with the exception that if an object is added or updated the first DAG build will be scheduled.

I think rescheduling the DAG build until the full cache sync is the only solution.

Reply:
Do you know what would be then the reason for `HasSynced()` to return `false` without it implying also that there is still a resource in the initial list that will eventually be delivered to the handler? Now it seems one cannot use `HasSynced()` call within event handlers `onAdd` but instead it _must_ be polled with an outside trigger.

Reply:
I have some gap in my thinking so I must ask: what are the other handlers in context of the client-go library running inside Contour process? 😅

Reply:
I see for example `ServiceStatusLoadBalancerWatcher` is a handler whose `HasSynced` is included in the checklist. (And I think it's the only one)

Reply:
You know, we can filter on the handler type and only check for contour-handler-related `HasSynced` functions. But that makes the code logic overly complicated which I think is not worth it.

There are handlers that wrap the contour handler and their `HasSynced` function should be accounted for. This requires keeping a list of handlers that should be looked after and then filtering on that list.

If you think the complication is worthwhile, I'm okay (and happy😁) to implement it.

Reply:
> I see for example ServiceStatusLoadBalancerWatcher is a handler whose HasSynced is included in the checklist. (And I think it's the only one)

Ah now I got you! With other handlers you meant other than `EventHandler` in `contour` package. I somehow thought you meant handlers that are not defined by contour at all. Sorry for my confusion 🤣

> You know, we can filter on the handler type and only check for contour-handler-related HasSynced functions.

Not worth it! The hang remains.

To test the theory where one of the other handlers without `Start()` & `Finished()` causes the hang, I put a parameter to `informOnResource()` allowing to skip adding `HasSynced()` methods for those.  I skipped it for `ServiceStatusLoadBalancerWatcher` and `EndpointsTranslator`. The hang was still there.

I wanted to see which of the `HasSynced()` functions is the last to turn `true` in a run that ended up in a hang. I added printouts to the return value of each separate `HasSynced()` and a timer that polls `SingleFileTracker.HasSynced()` to see all the statuses also after the last initial resource was already delivered to us. It seems to be always `Secret` which can stay `false` after last `OnAdd` is called and turn `true` only sometime later. 

For a moment I was thinking maybe `contour.EventRecorder` wrapper causes problems, since its `Start()` is not called immediately at `contour.EventRecorder.OnAdd` but only later when the wrapper calls next handler `contour.EventHandler.OnAdd`. But it should not matter when `Start()` is called. The whole call chain is synchronous so any moment is OK.

When I took a brief look of relevant parts of client-go to understand this PR better, I thought the status is just  combination of:
* counting `Start()` & `Finished()` call pairs
* the size of to-be-delivered initial resources

But it now seems there is more in it, so I apologize for requesting for the removal of the polling. You got right initially!

Reply:
Never mind. Thank you for your thorough review.